[DCP] Cache save plans: planner helpers and interface updates

saumishr opened the pull request:
Summary:
This PR updates the planner interface and introduces the class variables to cache the local and global plans. 
Two new helpers are also introduced which will be used to compare if the plans have changed across save attempts and merge the delta plans.

Test Plan: UTs

Differential Revision: D69224488




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @ekr0

meetv18 reviewed (commented):
Overall LGTM, just looking for clarification on the final_save_plan v/s save_plan comment.

saumishr commented on torch/distributed/checkpoint/planner.py:
```diff
@@ -203,6 +203,11 @@ class SavePlanner(abc.ABC):
     >>>         return global_plan, metadata
     """
 
+    _cached_save_plan: Dict[str, SavePlan] = {}
+    _cached_final_save_plan: Dict[str, SavePlan] = {}
+    _cached_all_plans: Dict[str, list[SavePlan]] = {}
+    _cached_global_plan: Dict[str, list[SavePlan]] = {}
+
```
Yes. Let me add that now. _cached_save_plan is the local plan produced by the _create_local_plan. However this plan can change after de-dupe or during global plan creation due to additional metadata. Hence we can not use that to send to the write_data method. So basically, `_cached_save_plan` is the local plan and _cached_final_save_plan is the plan after de-dupe and any additional metadata during global_plan creation.

saumishr committed:
[DCP] Cache save plans: planner helpers and interface updates (#147116)

Summary:

This PR updates the planner interface and introduces the class variables to cache the local and global plans. 
Two new helpers are also introduced which will be used to compare if the plans have changed across save attempts and merge the delta plans.

Test Plan: UTs

Reviewed By: MeetVadakkanchery

Differential Revision: D69224488

huydhn commented:
@pytorchbot merge -f 'The macos failure might be related but I am not sure'

As this PR has been landed internally, I will merge it.  But if the failure shows up in trunk, it will be reverted
